toolchain/lex/tokenized_buffer.cpp

```diff
@@ -270,252 +271,266 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
-        }
+    if (literal->is_multi_line()) {
+      while (current_line_info()->start + current_line_info()->length <
+             position) {
+        ++line_index_;
+        current_line_info()->indent = string_column;
       }
+      // Note that we've updated the current line at this point, but
+      // `set_indent_` is already true from above. That remains correct as the
+      // last line of the multi-line literal *also* has its indent set.
     }
 
     if (literal->is_terminated()) {
       auto token =
-          buffer_->AddToken({.kind = TokenKind::StringLiteral,
-                             .token_line = string_line,
-                             .column = string_column,
-                             .literal_index = static_cast<int32_t>(
-                                 buffer_->literal_string_storage_.size())});
-      buffer_->literal_string_storage_.push_back(
+          buffer_.AddToken({.kind = TokenKind::StringLiteral,
+                            .token_line = string_line,
+                            .column = string_column,
+                            .literal_index = static_cast<int32_t>(
+                                buffer_.literal_string_storage_.size())});
+      buffer_.literal_string_storage_.push_back(
           literal->ComputeValue(emitter_));
       return token;
     } else {
       CARBON_DIAGNOSTIC(UnterminatedString, Error,
                         "String is missing a terminator.");
       emitter_.Emit(literal->text().begin(), UnterminatedString);
-      return buffer_->AddToken({.kind = TokenKind::Error,
-                                .token_line = string_line,
-                                .column = string_column,
-                                .error_length = literal_size});
+      return buffer_.AddToken(
+          {.kind = TokenKind::Error,
+           .token_line = string_line,
+           .column = string_column,
+           .error_length = static_cast<int32_t>(literal_size)});
     }
   }
 
-  auto LexOneCharSymbolToken(llvm::StringRef& source_text, TokenKind kind)
-      -> Token {
+  auto LexOneCharSymbolToken(llvm::StringRef source_text, TokenKind kind,
+                             ssize_t& position) -> Token {
     // Verify in a debug build that the incoming token kind is correct.
     CARBON_DCHECK(kind != TokenKind::Error);
     CARBON_DCHECK(kind.fixed_spelling().size() == 1);
-    CARBON_DCHECK(source_text.front() == kind.fixed_spelling().front())
-        << "Source text starts with '" << source_text.front()
+    CARBON_DCHECK(source_text[position] == kind.fixed_spelling().front())
+        << "Source text starts with '" << source_text[position]
         << "' instead of the spelling '" << kind.fixed_spelling()
         << "' of the incoming token kind '" << kind << "'";
 
+    int column = ComputeColumn(position);
     if (!set_indent_) {
-      current_line_info_->indent = current_column_;
+      current_line_info()->indent = column;
       set_indent_ = true;
     }
```

Comment:
Can we factor this out into a function? We're doing this same work in a lot of places.

Hm, maybe `LexVerticalWhitespace` should call `LexHorizontalWhitespace` and then set the indent (and `LexStartOfFile` should do the same), and then we can remove the indent tracking from all the other handlers, other than the special case for multi-line string literals?

Reply:
I had been thinking similarly about factoring it out, but I *really* like your other idea of building this into the whitespace scanning rather than everything else. I'll work on that in a follow-up. I suspect that'll be a performance win in its own right.